[event-hubs] improve EventDataBatch.tryAdd performance

chradek opened the pull request:
Fixes #8567 

Previously, the `tryAdd` method would encode an AMQP message that contained all the events added to the batch in order to calculate the size of the batch with the new event added. This ends up being very expensive as more events are added to the batch (Making adding an event an O(n^2) operation).

With this change, the individual events are still encoded and stored as before, but now we calculate what the new size will be based on the known overhead of creating a batch. I took the overhead values from what [.NET](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventBatch.cs#L23-L29) and python have (5 bytes for small messages, 8 bytes for large messages) and verified through testing that these are correct. We also create an empty batch (with message annotations if applicable) the 1st time a message is able to be added to the batch in order to calculate the overhead for the batch envelope.

I tested these changes against the current implementation a few different ways.
1. I created 2 batches (one using the old tryAdd, one using the new tryAdd), each with 10,000 identical events. I found that:
   - The `_sizeInBytes` and event counts matched exactly in both.
   - The total time to add events dropped drastically (from 30404 ms to 68 ms in this test).
2. I created 2 batches (one using the old tryAdd, one using the new tryAdd) each with 5000 random events that ranged in size of 1 byte to 512 bytes. Again the batch sizes matched.

Since this was a perf improvement with no API changes, I didn't add many additional tests. The existing tests continue to pass. I think this is a good operation to write performance tests around so we don't regress in future updates but that might be better in a separate PR.

richardpark-msft commented on sdk/eventhub/event-hubs/src/eventDataBatch.ts:
```diff
@@ -258,42 +288,50 @@ export class EventDataBatchImpl implements EventDataBatch {
-    const currentSize = encodedBatchMessage.length;
+    const messageSize = encodedMessage.length;
+    const messageOverhead =
+      messageSize <= smallMessageMaxBytes ? smallMessageOverhead : largeMessageOverhead;
```
I can't respect you anymore. Where's the mysterious array with just two constants in it!?

chradek commented on sdk/eventhub/event-hubs/review/event-hubs.api.md:
```diff
@@ -59,9 +59,9 @@ export interface EventData {
-    readonly maxSizeInBytes: number;
     // @internal
-    readonly _message: Buffer | undefined;
+    _generateMessage(): Buffer;
```
This looks like a scarier change than it is. Note that it is marked internal and does not appear in the type definitions (neither did `_message`). This was only ever used internally.

chradek replied:
/cc @bterlson , I want to make sure you're ok with this change since it shows up in the api review file.

chradek commented:
Renamed the `_message` getter to `_generateMessage` so future devs are more aware that this method does some work.

Also added a test to make sure events are encoded properly when user properties are added.
